leshan-client-cf/src/main/java/org/eclipse/leshan/client/californium/ObserveCompositeRelationFilter.java

```diff
@@ -0,0 +1,95 @@
+ * and the Eclipse Distribution License is available at
+ *    http://www.eclipse.org/org/documents/edl-v10.html.
+ *
+ * Contributors:
+ *     Michał Wadowski (Orange) - Add Observe-Composite feature.
+ *******************************************************************************/
+package org.eclipse.leshan.client.californium;
+
+import java.util.Arrays;
+import java.util.List;
+
+import org.eclipse.californium.core.coap.CoAP;
+import org.eclipse.californium.core.coap.Request;
+import org.eclipse.californium.core.observe.ObserveRelation;
+import org.eclipse.californium.core.observe.ObserveRelationFilter;
+import org.eclipse.leshan.core.node.LwM2mPath;
+import org.eclipse.leshan.core.node.codec.CodecException;
+import org.eclipse.leshan.core.node.codec.LwM2mDecoder;
+import org.eclipse.leshan.core.request.ContentFormat;
+
+/**
+ * An {@link ObserveCompositeRelationFilter} which select {@link ObserveRelation} based on one of resource URIs.
+ */
+class ObserveCompositeRelationFilter implements ObserveRelationFilter {
+
+    private final LwM2mDecoder decoder;
+    private final List<LwM2mPath> paths;
+
+    /**
+     * Instantiates {@link ObserveCompositeRelationFilter} basing on resource URIs.
+     *
+     * @param decoder {@link LwM2mDecoder} which will decode data in supported content format.
+     * @param paths the list of {@link LwM2mPath}
+     */
+    public ObserveCompositeRelationFilter(LwM2mDecoder decoder, LwM2mPath... paths) {
+        this.decoder = decoder;
+        this.paths = Arrays.asList(paths);
+    }
+
+    @Override
+    public boolean accept(ObserveRelation relation) {
+        Request request = getRequest(relation);
+
+        if (!isValidObserveCompositeRequest(request)) {
+            return false;
+        }
+
+        List<LwM2mPath> observationPaths = getObserveRequestPaths(request);
+
+        if (observationPaths != null) {
+            for (LwM2mPath observePath : observationPaths) {
+                for (LwM2mPath path : paths) {
+                    if (path.startWith(observePath)) {
+                        return true;
+                    }
+                }
+            }
+        }
+
+        return false;
+    }
+
+    private boolean isValidObserveCompositeRequest(Request request) {
+        if (!request.getCode().equals(CoAP.Code.FETCH)) {
+            return false;
+        }
+        if (!request.getOptions().hasContentFormat()) {
+            return false;
+        }
+        return true;
+    }
+
+    private Request getRequest(ObserveRelation relation) {
+        return relation.getExchange().getRequest();
+    }
+
+    private List<LwM2mPath> getObserveRequestPaths(Request request) {
+        ContentFormat contentFormat = ContentFormat.fromCode(request.getOptions().getContentFormat());
+
+        try {
+            return decoder.decodePaths(request.getPayload(), contentFormat);
+        } catch (CodecException e) {
+            return null;
+        }
+    }
```

Comment:
I wonder if it's worth it to replace the decoder way and use  : 
 - `Request.setUserContext(Map<String, String>)` in RootResource. 
 - `Map<String, String> Request.getUserContext()` here.
 
 The idea is avoid to decode paths each time `ObserveRelationFilter` is called  but only first time we receive the request. 
 But the issue is that `UserContext` value string only and so we need to serialize / de-serialize it anyway.
 
 Or we need a way to store more than String at californium level.
 But If I remember well String was chosen to make serialization easier. 
 
 @boaks  maybe you have an idea for this "issue" ? (I can give more context if needed)

Reply:
I created PR #1084 witch tries to solve this issue. I think that simple splitting values by character will end up with less overhead.